File: tools/bpm.js
#!/usr/local/bin/node

"use strict";

var fs = require('fs');

if (process.argv.length < 3) {
  console.log("Usage:");
  console.log("  " + process.argv[1] + " " + "file_analysis.json");
  console.log("Prints the bpm of an echonest analysis file.");
  process.exit(0);
}


function mean(type, v) {

  var total = 0.0;
  for (var i = 1; i < v.length - 1; i++) {
    total += v[i+1][0]- v[i][0];
    console.log(v[i+1][0]);
    console.log(60/(v[i+1][0]- v[i][0]));
  }

  var meanBeatLength = total / (v.length - 2);
  var meanBeatTempo = 60. / meanBeatLength;

   console.log("manual BPM " + Math.round(meanBeatTempo*10)/10 + "");
}

try {
  var file_name = process.argv[2];
  var analysis = JSON.parse(fs.readFileSync(file_name));

  if (typeof analysis === "object" && typeof analysis["track"] === "object") {
    console.log("with confidence " + analysis["track"]["tempo_confidence"] + ", BPM " + Math.round(analysis["track"]["tempo"]*10)/10 + "");
  }
  else {
    mean("beats", analysis)
}
}
catch (e) {
  console.log("BPM could not be determined.");
}

File: tools/bpm.py
#!/usr/bin/python

import os, argparse, math, json

parser = argparse.ArgumentParser(description='Calculate the BPM from a JSON file.')
parser.add_argument('file', action="store", metavar='file', type=argparse.FileType('r'))
parser.add_argument('-v', action="store_true", default=False, help="verbose")
args = parser.parse_args()


def mean(type, v):
  total = 0.0
  for i in range(1, len(v)-1):
    total += v[i+1][0]- v[i][0];
    if args.v:
      print v[i+1][0], 60/(v[i+1][0]- v[i][0])

  meanBeatLength = total / (len(v) - 2)
  meanBeatTempo = 60. / meanBeatLength

  print "manual BPM", round(meanBeatTempo, 1)

try:
  
  analysis = json.load(args.file)

  if type(analysis) == dict and type(analysis["track"]) == dict:
    confidence = analysis["track"]["tempo_confidence"]
    bpm = round(analysis["track"]["tempo"], 1)
    print "with confidence", confidence, ", BPM", bpm
  else:
    mean("beats", analysis)
except:
  print "BPM could not be determined."